traits/trait_list_object.py

```diff
@@ -53,465 +90,769 @@ def __init__(self, index=0, removed=None, added=None):
-    trait will be copied into a new TraitListObject instance.
-
-    Mutation of the TraitListObject will fire a "name_items" event with
-    appropriate index, added and removed values.  In the case of setting
-    or deleting items with a slice, the index will hold either:
-
-    - the location of the first element changed, if the step is 1.  In
-      this case the last element changed can be inferred from the
-      length of the removed values.
-    - the slice that was used, otherwise.
+class TraitList(list):
+    """ A subclass of list that validates and notifies listeners of changes.
 
     Parameters
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : HasTraits instance
-        The HasTraits instance that the list has been set as an attribute for.
-    name : str
-        The name of the attribute on the object.
     value : list
-        The list of values to initialize the TraitListObject with.
+        The value for the list
+    validator : callable, optional
+        Called to validate items in the list.
+        With the signature::
+
+        validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+        notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, no notification fired
 
     Attributes
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : weak reference to a HasTraits instance
-        A weak reference to a HasTraits instance that the list has been set
-        as an attribute for.
-    name : str
-        The name of the attribute on the object.
-    name_items : str
-        The name of the items event trait that the trait list will fire when
-        mutated.
+    validator : callable, optional
+        Called to validate items in the list
+        With the signature::
+
+        validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+        notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, no notification fired
+
     """
 
-    def __init__(self, trait, object, name, value):
-        self.trait = trait
-        self.object = ref(object)
-        self.name = name
-        self.name_items = None
-        if trait.has_items:
-            self.name_items = name + "_items"
+    def __new__(cls, *args, **kwargs):
+        obj = super().__new__(cls)
+        obj.validator = kwargs.get('validator', None)
```

Comment:
What we could do here is set `obj.validator` to `None` and `obj.notifiers` to the empty list regardless of the values of the keywords: we don't need to do more than ensure the object has a consistent state in this method. Then we can use `__getnewargs__` instead of `__getnewargs_ex__`, so  that we stay  compatible with Python 3.5.

Reply:
(and to be clear, the `__init__` method would then be responsible for initialising the list of notifiers to the right thing).

By the way, do we have any actual use-cases for creating a `TraitList` with a predetermined collection of notifiers, or will the notifiers always be empty at creation time?

Reply:
I can't think of any, we can just do away with `__getnewargs__`, will reintroduce it if anyone thinks its needed.